### PR TITLE
Fix logei_candidates_func in `botorch.py`

### PR DESCRIPTION
**Pull request by sousu4**
## Motivation

#4961 resolved.
I made logei_candidates_func compatible with constraints and added corresponding tests in tests/integration_tests/test_botorch.py.

## Description of the changes

I have made modifications to logei_candidates_func in optuna/integration/botorch.py and test_botorch_specify_candidates_func in tests/integration_tests/test_botorch.py.

### CODE REVIEW

File: optuna/integration/botorch.py [hunk 1]
```diff
@@ -119,11 +120,24 @@ def logei_candidates_func(
 
     if train_obj.size(-1) != 1:
         raise ValueError("Objective may only contain single values with logEI.")
+
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:123]: This line addition is unnecessary. Could you revert this?

File: optuna/integration/botorch.py [hunk 2]
```diff
@@ -119,11 +120,24 @@ def logei_candidates_func(
...
-            "Constraint is not supported with logei_candidates_func. "
-            + "Please use qei_candidates_func instead."
-        )
+        train_y = torch.cat([train_obj, train_con], dim=-1)
+
+        is_feas = (train_con <= 0).all(dim=-1)
+        train_obj_feas = train_obj[is_feas]
+
+        if train_obj_feas.numel() == 0:
+            # TODO(hvy): Do not use 0 as the best observation.
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:131]: The author of this TODO comment should be you.

File: optuna/integration/botorch.py [hunk 3]
```diff
@@ -119,11 +120,24 @@ def logei_candidates_func(
...
+            _logger.warning(
+                "No objective values are feasible. Using 0 as the best objective in qEI."
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:133]: The acquisition function is not `qEI` but `LogConstrainedExpectedImprovement`.

File: optuna/integration/botorch.py [hunk 4]
```diff
@@ -119,11 +120,24 @@ def logei_candidates_func(
...
+            )
+            best_f = torch.zeros(())
+        else:
+            best_f = train_obj_feas.max()
+
+        n_constraints = train_con.size(1)
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:139]: This variable should be defined near the use of it. How about moving this to right before the line 152?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have several comments. PTAL.

**Comment by HideakiImamura** [comment]
The docstring in the line 82 and 193 should be fixed. In addition, the line 621 should be fixed to use `logei_candidates_func` for constrained problems.

**Comment by HideakiImamura** [comment]
I conducted the constrained optimization problem with the implemented `logei_candidates_func` and the existing `qei_candidates_func` by using the following code snippets, and got the following figures. Both candidates function works fine, but none of them performs well (I mean both algorithms search the infeasible region at most). This should be discussed after this PR.

```python
import optuna


def objective(trial):
    x = trial.suggest_float('x', 0, 4)

    constraint = - x + 3
    trial.set_user_attr('constraint', [constraint])

    return (x - 2) ** 2


def constraint(trial):
    return trial.user_attrs['constraint']


if __name__ == "__main__":
    study = optuna.create_study(
        sampler=optuna.integration.BoTorchSampler(
            candidates_func=optuna.integration.botorch.logei_candidates_func,
            constraints_func=constraint,
        ),
        storage='sqlite:///example.db',
        study_name='logei',
        load_if_exists=True,
    )
    study.optimize(objective, n_trials=100)

    optuna.visualization.plot_slice(study).show()
    optuna.visualization.plot_optimization_history(study).show()
```

| Candidates function | Slice plot | History plot |
| --- | --- | --- |
| `logei_candidates_func` | newplot (45) | newplot (46) |
| `qei_candidates_func` | newplot (47) | newplot (48) |

File: optuna/integration/botorch.py [hunk 5]
```diff
@@ -168,7 +190,7 @@ def qei_candidates_func(
     """Quasi MC-based batch Expected Improvement (qEI).
 
     The default value of ``candidates_func`` in :class:`~optuna.integration.BoTorchSampler`
-    with single-objective optimization for constrained problems.
+    with single-objective optimization.
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:193]: This content is confusing. I think we should remove this. 

In addition, we need to fix the getter function of the candidates function.

File: optuna/integration/botorch.py [hunk 6]
```diff
@@ -120,10 +123,21 @@ def logei_candidates_func(
     if train_obj.size(-1) != 1:
         raise ValueError("Objective may only contain single values with logEI.")
     if train_con is not None:
-        raise ValueError(
-            "Constraint is not supported with logei_candidates_func. "
-            + "Please use qei_candidates_func instead."
-        )
+        n_constraints = train_con.size(1)
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:126]: How about the following?

```python
n_constraints = train_con.size(1) if train_con is not None elase 0
if n_constraints > 0:
    ...
```

File: optuna/integration/botorch.py [hunk 7]
```diff
@@ -133,11 +147,18 @@ def logei_candidates_func(
     model = SingleTaskGP(train_x, train_y, outcome_transform=Standardize(m=train_y.size(-1)))
     mll = ExactMarginalLogLikelihood(model.likelihood, model)
     fit_gpytorch_mll(mll)
-
-    acqf = LogExpectedImprovement(
-        model=model,
-        best_f=best_f,
-    )
+    if train_con is not None and n_constraints > 0:
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:150]:
```suggestion
    if n_constraints > 0:
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the update. I have two minor comments.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update! LGTM!

**Comment by HideakiImamura** [comment]
@contramundum53 Could you review this PR?

File: optuna/integration/botorch.py [hunk 8]
```diff
@@ -119,11 +122,23 @@ def logei_candidates_func(
...
+
+        is_feas = (train_con <= 0).all(dim=-1)
+        train_obj_feas = train_obj[is_feas]
+
+        if train_obj_feas.numel() == 0:
+            # TODO(sousu4): Do not use 0 as the best observation.
+            _logger.warning(
+                "No objective values are feasible. Using 0 as the best objective in logEI."
+            )
+            best_f = torch.zeros(())
```
Comment (contramundum53) [inline optuna/integration/botorch.py:138]: How about using the minimum value of `train_obj`?
```suggestion
            best_f = train_obj.min()
```

File: optuna/integration/botorch.py [hunk 9]
```diff
@@ -133,11 +148,18 @@ def logei_candidates_func(
...
-    acqf = LogExpectedImprovement(
-        model=model,
-        best_f=best_f,
-    )
+    if n_constraints > 0:
+        acqf = LogConstrainedExpectedImprovement(
+            model=model,
+            best_f=best_f,
+            objective_index=0,
+            constraints={i: (None, None) for i in range(1, n_constraints + 1)},
```
Comment (contramundum53) [inline optuna/integration/botorch.py:156]: I think you need to specify the constraint bound (currently no bound is specified).
```suggestion
            constraints={i: (None, 0.0) for i in range(1, n_constraints + 1)},
```

Comment (sousu4) [reply optuna/integration/botorch.py]: I have incorporated all of your comments. Please review them again.

**Comment by contramundum53** [comment]
Sorry, I was off the last week. Let me review this PR again.

**Review by contramundum53 — APPROVED** [review]
LGTM!

**Comment by contramundum53** [comment]
Maybe we should add a test for constrained optimization. That can be a follow-up, however.

File: optuna/integration/botorch.py [hunk 10]
```diff
@@ -119,11 +122,23 @@ def logei_candidates_func(
...
+    n_constraints = train_con.size(1) if train_con is not None else 0
+    if n_constraints > 0:
+        assert train_con is not None
+        train_y = torch.cat([train_obj, train_con], dim=-1)
+
+        is_feas = (train_con <= 0).all(dim=-1)
+        train_obj_feas = train_obj[is_feas]
+
+        if train_obj_feas.numel() == 0:
+            # TODO(sousu4): Do not use 0 as the best observation.
```
Comment (contramundum53) [inline optuna/integration/botorch.py:134]:
```suggestion
```
Could you remove this comment?

**Comment by contramundum53** [comment]
The code works for me. Other than the above comment, LGTM!
```
#%%
import optuna
from optuna.integration import BoTorchSampler

study = optuna.create_study(sampler=BoTorchSampler(
    constraints_func=lambda t: [t.params["x"] + 1, t.params["x"] + t.params["y"] + 3]
))

def objective(trial):
    x = trial.suggest_float("x", -3, 3)
    y = trial.suggest_float("y", -3, 3)

    return x ** 2 + y ** 2


study.optimize(objective, n_trials=100)
#%%
from optuna.visualization import plot_rank
plot_rank(study)
```

image

**Review by contramundum53 — APPROVED** [review]
LGTM! Thanks for your contribution!

**Pull request merged**